### DISCUSSION TITLE
Hierarchical time series

### DISCUSSION THREAD

**Discussion by TheKeyblader** [ravendb/ravendb#15110] [Q&A]
Hello👋

I would like to discuss a scenario where I can't find the most effective method.

we have multiple site with an timeseries who contains the number of person has enter and exit the site.
we have multiple client that contain multiple site.

here some classes to help visualize
```csharp
public record Site(string Id, string Name, string CustomerId);
```
```csharp
public record Customer(string Id, string Name);
```

so the question is how to group multiple time series in one ?
something like 
```
from 'Employees'
where clientId == 'customers/1'
select sum(timeseries(from 'HeartRates'))
```
# Solution 1
sum all time series on client side. (tested)

## Avantages
- support all group options
## Disavantages
- pretty heavy on network
- more site means longer query time

# Solution 2
use map reduce with artificial documents (tested)

## Avantages
- query the data you need
## Disavantages
- heavy on storage
- only one group criteria per index
- complicated to manage

# Solution 3
use multiple incremental time series for each level (not tested)
but how to manage error like I need to remove one erroneous site from aggregate

## Avantages
- easy to query
- support all group options
## Disavantages
- heavy on storage but maybe less than artificial documents
- prone to bug or error
- complicated to manage revert or correct error

# Solution 4
maybe some magic trick with javascript index

Many thanks in advance

**Comment by ayende** [comment]
What are you trying to query?
Give me all the time series for `customers/1` in `sites/2` ?

**Reply by TheKeyblader** [reply]
The customer 1 has 2 site
sites/1 with time series named "inc:Count"
| Date             | Enter | Exit |
|------------------|-------|------|
| 2022/28/09 14:00 | 1     | 2    |
| 2022/28/09 15:00 | 3     | 7    |
| 2022/28/09 16:00 | 4     | 2    |

sites/2 with time series named "inc:Count"
| Date             | Enter | Exit |
|------------------|-------|------|
| 2022/28/09 14:00 | 1     | 2    |
| 2022/28/09 15:00 | 1     | 4    |
| 2022/28/09 16:00 | 9     | 7    |

I want the best way to get the sum of the 2 time series for ``customers/1`` 
something like 
| Date             | Enter | Exit |
|------------------|-------|------|
| 2022/28/09 14:00 | 2     | 4    |
| 2022/28/09 15:00 | 4     | 11    |
| 2022/28/09 16:00 | 13     | 9    |

**Reply by ayende** [reply]
Are they all on the same document? If so, you can merge them at query time, I believe.

You can do it like this:

```
declare timeseries f(name){
    from name
}
from "Employees" as e
select {
    F: [f("HeartRates"),f("HeartRates")]
}

```

And in the `{}` you can write the merge behavior

**Reply by TheKeyblader** [reply]
No they are in different documents

Here my actual solution I found:

Map Reduce index

Map
```csharp
timeSeries.Areas.Sales.Select(timeserie => new {
    timeserie = timeserie,
    area = this.LoadDocument(timeserie.DocumentId, "Areas")
}).SelectMany(this0 => this0.timeserie.Entries, (this0, entry) => new {
    Date = entry.Timestamp.Date,
    SiteId = this0.area.SiteId,
    Tickets = entry.Values[0],
    Amount = entry.Values[2],
    Items = entry.Values[1]
})
```
Reduce
```csharp
results.GroupBy(result => new {
    Date = result.Date,
    SiteId = result.SiteId
}).Select(g => new {
    Date = g.Key.Date,
    SiteId = g.Key.SiteId,
    Tickets = Enumerable.Sum(g, x => ((double) x.Tickets)),
    Amount = Enumerable.Sum(g, x0 => ((double) x0.Amount)),
    Items = Enumerable.Sum(g, x1 => ((double) x1.Items))
})
```
With Output Reduce Results to Collection set to DailySalesBySites
